dipy/io/stateful_tractogram.py

```diff
@@ -0,0 +1,798 @@
+
+        Returns
+        -------
+        output : ndarray
+            8 corners of the XYZ aligned box, all zeros if no streamlines
+        """
+        if self._tractogram.streamlines.data.size > 0:
+            bbox_min = np.min(self._tractogram.streamlines.data, axis=0)
+            bbox_max = np.max(self._tractogram.streamlines.data, axis=0)
+
+            return np.asarray(list(product(*zip(bbox_min, bbox_max))))
+
+        return np.zeros((8, 3))
+
+    def is_bbox_in_vox_valid(self):
+        """ Verify that the bounding box is valid in voxel space
+        Will transform the streamlines for OBB, slow for big tractogram
+
+        Returns
+        -------
+        output : bool
+            Are the streamlines within the volume of the associated reference
+        """
+        old_space = deepcopy(self.get_current_space())
+        old_shift = deepcopy(self.get_current_shift())
+
+        # Do to rotation, equivalent of a OBB must be done
+        self.to_vox()
+        self.to_corner()
+        bbox_corners = deepcopy(self.compute_bounding_box())
+
+        is_valid = True
+        if np.any(bbox_corners < 0):
+            logging.error('Voxel space values lower than 0.0')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if np.any(bbox_corners[:, 0] > self._dimensions[0]) or \
+                np.any(bbox_corners[:, 1] > self._dimensions[1]) or \
+                np.any(bbox_corners[:, 2] > self._dimensions[2]):
+            logging.error('Voxel space values higher than dimensions')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if old_space == Space.RASMM:
+            self.to_rasmm()
+        elif old_space == Space.VOXMM:
+            self.to_voxmm()
+
+        if not old_shift:
+            self.to_center()
+
+        return is_valid
+
+    def remove_invalid_streamlines(self):
+        old_space = deepcopy(self.get_current_space())
+        old_shift = deepcopy(self.get_current_shift())
+
+        self.to_vox()
+        self.to_corner()
+
+        min_condition = np.min(self._tractogram.streamlines.data,
+                               axis=1) < 0.0
+        max_condition = np.any(self._tractogram.streamlines.data >
+                               self._dimensions, axis=1)
+        ic_offsets_indices = np.where(np.logical_or(min_condition,
+                                                    max_condition))[0]
+
+        ic_indices = []
+        for i in ic_offsets_indices:
+            ic_indices.append(bisect(
+                self._tractogram.streamlines._offsets, i) - 1)
+
+        indices_to_save = np.setdiff1d(np.arange(len(self._tractogram)),
+                                       np.array(ic_indices)).astype(int)
+
+        tmp_streamlines = \
+            itemgetter(*indices_to_save)(self.get_streamlines_copy())
+        tmp_data_per_point = {}
+        tmp_data_per_streamline = {}
+
+        for key in self._tractogram.data_per_point:
```

Comment:
What are you trying to do here? What is the purpose?

Reply:
Considering this function removes streamlines that are invalid, I remove their data (per_point, per_streamline) so the new object is still consistent and valid for the `nibabel.tractogram` object